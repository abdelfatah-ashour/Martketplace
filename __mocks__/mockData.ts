export const mockCategories = [
  "electronics",
  "jewelery",
  "men's clothing",
  "women's clothing",
];

export const mockProducts = [
  {
    id: 1,
    title: "product 1",
    price: 200,
    category: "women",
    description:
      "great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.",
    image:
      "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
  },
  {
    id: 2,
    title: "product 2",
    price: 200,
    category: "men",
    description:
      "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
    image: "https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg",
  },
];
