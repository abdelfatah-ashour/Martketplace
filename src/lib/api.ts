import { config } from "./config";
const { apiUrl } = config;

export default class Api {
  /**
   * Performs a GET request to the specified URL.
   * @param {string} url - The URL to fetch data from.
   * @returns {Promise<T>} A promise that resolves to the fetched data of type T.
   * @template T - The type of data expected to be fetched.
   */
  async get<T>(url: string) {
    return await fetch(`${apiUrl}${url}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((res) => res.json())
      .then((res) => res as T)
      .catch(() => null);
  }

  /**
   * Performs a POST request to the specified URL with the given body.
   * @param {string} url - The URL to send the POST request to.
   * @param {Record<string, unknown>} body - The body of the POST request.
   * @returns {Promise<T>} A promise that resolves to the response data of type T.
   * @template T - The type of data expected in the response.
   */
  async post<T>(url: string, body: Record<string, unknown>) {
    return await fetch(`${apiUrl}${url}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    })
      .then((res) => res.json())
      .then((res) => res as T)
      .catch(() => null);
  }
}
