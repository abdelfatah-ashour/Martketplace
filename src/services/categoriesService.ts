import Api from "@/lib/api";
import { create } from "zustand";

interface ICategories {
  categories: string[];
  selectedCategory: string | null;
  getCategories: () => void;
  selectCategory: (category: string) => void;
  rest: () => void;
}

const categoriesService = create<ICategories>((set) => ({
  categories: [],
  selectedCategory: null,
  async getCategories() {
    const categories = await new Api().get<string[]>("/products/categories");

    if (categories?.length) {
      set(() => ({
        categories,
      }));
    }
  },
  selectCategory(category) {
    set(() => ({
      selectedCategory: category,
    }));
  },
  rest() {
    set(() => ({
      ...categoriesService.getInitialState(),
    }));
  },
}));

export default categoriesService;
