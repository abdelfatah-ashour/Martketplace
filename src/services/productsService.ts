import { IProduct } from "@/@types/Product";
import Api from "@/lib/api";
import { create } from "zustand";

interface IProducts {
  products: IProduct[];
  loaded: boolean;
  loading: boolean;
  error: string | null;
  getProducts: () => void;
  getProductsByCategory: (category: string) => void;
}

const productService = create<IProducts>((set) => ({
  products: [],
  loading: false,
  loaded: false,
  error: null,

  async getProducts() {
    set(() => ({
      loading: true,
      products: [],
    }));

    const products = await new Api()
      .get<IProduct[]>("/products")
      .catch(() => {
        set(() => ({
          error: "something went wrong.",
        }));
      })
      .finally(() => {
        set(() => ({
          loading: false,
        }));
      });

    if (products?.length) {
      set(() => ({
        products,
        loaded: true,
      }));
    }
  },
  async getProductsByCategory(category) {
    set(() => ({
      loading: true,
      products: [],
    }));
    const products = await new Api()
      .get<IProduct[]>(`/products/category/${category}`)
      .catch(() => {
        set(() => ({
          error: "something went wrong.",
        }));
      })
      .finally(() => {
        set(() => ({
          loading: false,
        }));
      });

    if (products) {
      set(() => ({
        products,
        loaded: true,
      }));
    }
  },
}));

export default productService;
