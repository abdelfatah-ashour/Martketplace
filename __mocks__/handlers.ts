// src/mocks/handlers.js
import { http, HttpResponse } from "msw";
import { config } from "../src/lib/config";
import { mockCategories } from "./mockData";

const { apiUrl } = config;

export const handlers = [
  http.get(`${apiUrl}/products/categories`, async () => {
    return HttpResponse.json(mockCategories, { status: 200 });
  }),

  http.get(`${apiUrl}/products`, async () => {
    return HttpResponse.json(mockCategories, { status: 200 });
  }),
];
