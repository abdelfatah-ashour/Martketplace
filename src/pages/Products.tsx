import CategoriesFilter from "@/components/CategoriesFilter";
import ProductsGrid from "@/components/ProductsGrid";
import categoriesService from "@/services/categoriesService";
import { useEffect } from "react";

export default function Products() {
  const { getCategories, rest } = categoriesService();

  useEffect(() => {
    getCategories();

    return () => {
      rest();
    };
  }, [getCategories, rest]);

  return (
    <>
      <CategoriesFilter />
      <ProductsGrid />
    </>
  );
}
