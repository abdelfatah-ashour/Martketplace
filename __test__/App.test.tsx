import {
  cleanup,
  fireEvent,
  render,
  renderHook,
  waitFor,
} from "@testing-library/react";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import React from "react";
import { mockCategories, mockProducts } from "../__mocks__/mockData";
import { config } from "../src/lib/config";
import Products from "../src/pages/Products";
import categoriesService from "../src/services/categoriesService";
import productsService from "../src/services/productsService";

const { apiUrl } = config;

afterAll(() => {
  cleanup();
});

const server = setupServer(
  http.get(`${apiUrl}/products`, () =>
    HttpResponse.json(mockProducts, { status: 200 })
  ),
  http.get(`${apiUrl}/products/categories`, () =>
    HttpResponse.json(mockCategories, { status: 200 })
  ),
  http.get(`${apiUrl}/products/category/${mockCategories[0]}`, () =>
    HttpResponse.json(mockCategories, { status: 200 })
  )
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("Products", () => {
  describe("Render", async () => {
    it("Categories filter must be render correctly.", () => {
      const elem = render(<Products />);
      expect(elem).toMatchSnapshot();
    });

    it("Categories filter must be contain heading 'Shop by category'", async () => {
      renderHook(productsService);
      const { getByRole } = await waitFor(() => render(<Products />));

      const elem = await getByRole("heading", { level: 2 }).innerHTML;
      expect(elem).toMatch(/shop by category/i);
    });

    it("Categories must be render correct", async () => {
      renderHook(categoriesService);

      const { getAllByRole } = await waitFor(() => render(<Products />));

      const buttons = getAllByRole("button").length;
      expect(buttons).toBeTruthy();
    });

    it("Products must be render correct", async () => {
      renderHook(productsService);

      const { getAllByRole } = await waitFor(() => render(<Products />));

      const buttons = getAllByRole("button").length;
      expect(buttons).toBeTruthy();
    });
  });

  describe("Behaviors", async () => {
    it("Fetch and render products when select a category", async () => {
      renderHook(productsService);
      const { findAllByRole } = await waitFor(() => render(<Products />));

      const btnCategories = await findAllByRole("button");

      fireEvent.click(btnCategories[0]);

      const products = await waitFor(
        () => findAllByRole("heading", { level: 3 }),
        {
          timeout: 3000,
        }
      );

      expect(products.length).toBeTruthy();
    });
  });
});
