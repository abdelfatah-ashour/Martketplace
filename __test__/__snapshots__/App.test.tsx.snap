// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products > Render > Categories filter must be render correctly. 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full grid grid-cols-12 items-start my-10 gap-6 lg:gap-2"
      >
        <div
          class="col-span-12 lg:col-span-4"
        >
          <h2
            class="font-semibold"
          >
            Shop by category
          </h2>
          <p
            class="text-sm text-gray-500"
          >
            Discover the perfect products for your needs.
          </p>
        </div>
        <div
          class="col-span-12 lg:col-span-8 flex gap-3 lg:gap-6 lg:justify-end items-start"
        />
      </div>
      <div
        class="grid grid-cols-12 gap-6 gap-y-12"
      >
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="col-span-12 md:col-span-4 lg:col-span-3"
        >
          <div
            class="w-full"
          >
            <div
              class="relative py-3 flex flex-col gap-2 animate-pulse"
            >
              <div
                class="w-full h-64 rounded-lg bg-gray-300"
              />
              <div
                class="flex flex-col gap-2 animate-pulse"
              >
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
                <div
                  class="h-8 w-full bg-gray-300 rounded"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full grid grid-cols-12 items-start my-10 gap-6 lg:gap-2"
    >
      <div
        class="col-span-12 lg:col-span-4"
      >
        <h2
          class="font-semibold"
        >
          Shop by category
        </h2>
        <p
          class="text-sm text-gray-500"
        >
          Discover the perfect products for your needs.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-8 flex gap-3 lg:gap-6 lg:justify-end items-start"
      />
    </div>
    <div
      class="grid grid-cols-12 gap-6 gap-y-12"
    >
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4 lg:col-span-3"
      >
        <div
          class="w-full"
        >
          <div
            class="relative py-3 flex flex-col gap-2 animate-pulse"
          >
            <div
              class="w-full h-64 rounded-lg bg-gray-300"
            />
            <div
              class="flex flex-col gap-2 animate-pulse"
            >
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
              <div
                class="h-8 w-full bg-gray-300 rounded"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
